### PR TITLE
[C] FlexLayout, with managed engine

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

Add a new FlexLayout to Xamarin.Forms.

#### Credits

The Xamarin.Flex engine is a port of @lrz's native https://github.com/xamarin/flex.

Thanks to @rmarinho for the initial proof of concept and the almost-extensive test suite.

#### Complexity

Measure and Layout are O(n) and doesn't get more complex if you nest FlexLayout into each other. Layout and Measure are single-pass, top-down.

### Bugs Fixed ###

/

### API Changes ###

Added:
```csharp
public class Xamarin.Forms.FlexLayout : Layout<View>
{
	//BP, applies to the layout
	public static readonly BindableProperty DirectionProperty;
	public static readonly BindableProperty JustifyContentProperty;
	public static readonly BindableProperty AlignContentProperty;
	public static readonly BindableProperty AlignItemsProperty;
	public static readonly BindableProperty PositionProperty;
	public static readonly BindableProperty WrapProperty;

	//Attached ones, applies to children
	public static readonly BindableProperty OrderProperty;
	public static readonly BindableProperty GrowProperty;
	public static readonly BindableProperty ShrinkProperty;
	public static readonly BindableProperty AlignSelfProperty;
	public static readonly BindableProperty BasisProperty;

	//property accessors
	public FlexDirection Direction { get; set; }
	public FlexJustify JustifyContent { get; set; }
	public FlexAlignContent AlignContent { get; set; }
	public FlexAlignItems AlignItems { get; set; }
	public FlexPosition Position { get; set; }
	public FlexWrap Wrap { get; set; }

	public static int GetOrder(BindableObject bindable);
	public static void SetOrder(BindableObject bindable, int value);
	public static float GetGrow(BindableObject bindable);
	public static void SetGrow(BindableObject bindable, float value);
	public static float GetShrink(BindableObject bindable);
	public static void SetShrink(BindableObject bindable, float value);
	public static FlexAlignSelf GetAlignSelf(BindableObject bindable);
	public static void SetAlignSelf(BindableObject bindable, FlexAlignSelf value);
	public static FlexBasis GetBasis(BindableObject bindable);
	public static void SetBasis(BindableObject bindable, FlexBasis value);
}

public enum Xamarin.Forms.FlexJustify
{
	Center,
	Start,
	End,
	SpaceBetween,
	SpaceAround,
	SpaceEvenly,
}

public enum Xamarin.Forms.FlexPosition
{
	Relative,
	Absolute,
}

public enum Xamarin.Forms.FlexDirection
{
	Row,
	RowReverse,
	Column,
	ColumnReverse,
}

public enum Xamarin.Forms.FlexAlignContent
{
	Stretch,
	Center,
	Start,
	End,
	SpaceBetween,
	SpaceAround,
	SpaceEvenly,
}

public enum Xamarin.Forms.FlexAlignItems
{
	Stretch,
	Center,
	Start,
	End,
}

public enum Xamarin.Forms.FlexAlignSelf
{
	Auto,
	Stretch,
	Center,
	Start,
	End,
}

public enum Xamarin.Forms.FlexWrap
{
	NoWrap,
	Wrap,
	Reverse,
}

public struct Xamarin.Forms.FlexBasis
{
	public static FlexBasis Auto;
	public float Length { get; }
	public FlexBasis (float length, bool isRelative = false);
	public static implicit operator FlexBasis(float length);
}
```

### Behavioral Changes ###

/

### Working samples

https://gist.github.com/StephaneDelcroix/da29dcbb1a1f7501bab387e76ad0be17

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/FlexEnums.cs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	}
+
+	[TypeConverter(typeof(FlexAlignItemsTypeConverter))]
+	public enum FlexAlignItems
+	{
+		Stretch = Flex.Align.Stretch,
+		Center = Flex.Align.Center,
+		FlexStart = Flex.Align.Start,
+		FlexEnd = Flex.Align.End,
+		//Baseline = Flex.Align.Baseline,
```
Comment (hartez) [inline Xamarin.Forms.Core/FlexEnums.cs]: Was this intentionally removed?

File: Xamarin.Forms.Core/FlexEnums.cs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+	[TypeConverter(typeof(FlexAlignSelfTypeConverter))]
+	public enum FlexAlignSelf
+	{
+		Auto = Flex.Align.Auto,
+		Stretch = Flex.Align.Stretch,
+		Center = Flex.Align.Center,
+		FlexStart = Flex.Align.Start,
+		FlexEnd = Flex.Align.End,
+		//Baseline = Flex.Align.Baseline,
```
Comment (hartez) [inline Xamarin.Forms.Core/FlexEnums.cs]: Same here - intentionally removed?

File: Xamarin.Forms.Core/FlexLayout.cs [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+			item.Direction = (Flex.Direction)(FlexDirection)values[2];
+			item.JustifyContent = (Flex.Align)(FlexJustify)values[3];
+			item.Wrap = (Flex.Wrap)(FlexWrap)values[4];
+		}
+
+		void ClearLayout()
+		{
+			foreach (var child in Children)
+				RemoveChild(child);
+			//_root.Dispose();
```
Comment (hartez) [inline Xamarin.Forms.Core/FlexLayout.cs]: Is this intentional?

File: Xamarin.Forms.Core/FlexLayout.cs [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+			item.Height = (double)values[7] < 0 ? float.NaN : (float)(double)values[7];
+		}
+
+		void RemoveChild(View view)
+		{
+			if (_root == null)
+				return;
+			var item = GetFlexItem(view);
+			_root.Remove(item);
+			//item.Dispose();
```
Comment (hartez) [inline Xamarin.Forms.Core/FlexLayout.cs]: Intentional?

File: Xamarin.Forms.Core/FlexEnums.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/FlexEnums.cs]: yes, not supported by the engine

File: Xamarin.Forms.Core/FlexLayout.cs (hunk 3 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/FlexLayout.cs]: it should go, the managed engine doesn't need to be disposed.

File: Xamarin.Forms.Core/FlexLayout.cs (hunk 4 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/FlexLayout.cs]: same

File: Xamarin.Forms.Core/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -59,4 +59,17 @@
 [assembly: StyleProperty("padding-bottom", typeof(IPaddingElement), nameof(PaddingElement.PaddingBottomProperty), PropertyOwnerType = typeof(PaddingElement))]
 [assembly: StyleProperty("text-align", typeof(ITextAlignmentElement), nameof(TextAlignmentElement.HorizontalTextAlignmentProperty))]
 [assembly: StyleProperty("visibility", typeof(VisualElement), nameof(VisualElement.IsVisibleProperty))]
-[assembly: StyleProperty("width", typeof(VisualElement), nameof(VisualElement.WidthRequestProperty))]
\ No newline at end of file
+[assembly: StyleProperty("width", typeof(VisualElement), nameof(VisualElement.WidthRequestProperty))]
+
+//flex
+[assembly: StyleProperty("align-content", typeof(FlexLayout), nameof(FlexLayout.AlignContentProperty))]
+[assembly: StyleProperty("align-items", typeof(FlexLayout), nameof(FlexLayout.AlignSelfProperty))]
```
Comment (samhouts) [inline Xamarin.Forms.Core/Properties/AssemblyInfo.cs]: `AlignItemsProperty`?

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Properties/AssemblyInfo.cs]: good catch, thx

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/Properties/AssemblyInfo.cs]: fixed

File: Xamarin.Forms.Core/FlexLayout.cs (hunk 3 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/FlexLayout.cs]: fixed

File: Xamarin.Forms.Core/FlexEnums.cs (hunk 1 above)
Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/FlexEnums.cs]: left it commented out, so we know where we differ from the css spec

**Pull request merged**